src/lib/Simplify.hs

```diff
@@ -47,21 +49,51 @@ data WillLinearize = WillLinearize | NoLinearize
 -- them, and we try to preserve their identity for as long as we can. Since the only
 -- elimination form for function is application, we do some extra handling in simplifyApp.
 newtype SimplifyM (i::S) (o::S) (a:: *) = SimplifyM
-  { runSimplifyM' :: SubstReaderT AtomSubstVal (BuilderT (ReaderT WillLinearize HardFailM)) i o a }
+  { runSimplifyM'
+    :: SubstReaderT AtomSubstVal (DoubleBuilderT  (ReaderT WillLinearize HardFailM)) i o a }
   deriving ( Functor, Applicative, Monad, ScopeReader, EnvExtender, Fallible
-           , Builder, EnvReader, SubstReader AtomSubstVal, MonadFail )
+           , EnvReader, SubstReader AtomSubstVal, MonadFail
+           , Builder, HoistingTopBuilder)
 
-liftSimplifyM :: (SinkableE e, EnvReader m) => SimplifyM n n (e n) -> m n (e n)
+liftSimplifyM
+  :: (SinkableE e, SubstE Name e, TopBuilder m, Mut n)
+  => (forall l. DExt n l => SimplifyM n l (e l))
+  -> m n (e n)
 liftSimplifyM cont = do
-  liftM (runHardFail . flip runReaderT NoLinearize) $ liftBuilderT $ runSubstReaderT idSubst $ runSimplifyM' cont
+  Abs envFrag e <- liftM (runHardFail . flip runReaderT NoLinearize) $
+    liftDoubleBuilderT $ runSubstReaderT (sink idSubst) $ runSimplifyM' cont
+  emitEnv $ Abs envFrag e
 {-# INLINE liftSimplifyM #-}
 
+liftSimplifyMAssumeNoTopEmissions
+  :: (SinkableE e, SubstE Name e, EnvReader m)
+  => (forall l. DExt n l => SimplifyM l l (e l))
+  -> m n (e n)
+liftSimplifyMAssumeNoTopEmissions cont =
+  liftM (runHardFail . flip runReaderT NoLinearize) $
+    liftDoubleBuilderTAssumeNoTopEmissions $
+      runSubstReaderT (sink idSubst) $ runSimplifyM' cont
+{-# INLINE liftSimplifyMAssumeNoTopEmissions #-}
+
+liftDoubleBuilderTAssumeNoTopEmissions
```

Comment:
It's kind of a hack because of the way Imp calls into Simplify, which we're hoping to fix soon, so I didn't want to dignify it by putting it in Builder. I'll add a comment to that effect.

Reply:
done